tools/rewriter/DetermineAbi.cpp

```diff
@@ -157,18 +168,22 @@ abiSlotsForArg(const clang::QualType &qt,
         dyn_cast<llvm::StructType>(argInfo.getCoerceToType());
     if (STy) {
       if (argInfo.getCanBeFlattened()) {
+        // A struct is "flattenable" if its individual elements can be passed as arguments.
+        // In this case we classify each element of the struct and add each to the list.
         auto elems = STy->elements();
         std::vector<CAbiArgKind> out = {};
         for (const auto &elem : elems) {
           out.push_back(classifyLlvmType(*elem));
         }
         return out;
       } else {
+        // TODO why a single int slot when not flattenable?
         return {CAbiArgKind::Integral}; // one pointer in register, see `clang's
                                         // ClangToLLVMArgMapping::construct`
```

Comment:
As indicated by the comment below, non-flattenable types passed as `Direct` are a single pointer in a register. See the branch on `AI.isDirect() && AI.getCanBeFlattened() && STy` [here](https://clang.llvm.org/doxygen/CGCall_8cpp_source.html).